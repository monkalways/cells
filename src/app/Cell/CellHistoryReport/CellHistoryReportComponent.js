import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { withStyles } from '@material-ui/core/styles';
import {
  AppBar,
  BottomNavigation,
  BottomNavigationAction,
  Grid,
  Typography,
} from '@material-ui/core';
import ArrowBackIcon from '@material-ui/icons/ArrowBack';
import PrintIcon from '@material-ui/icons/Print';
import { withRouter } from 'react-router-dom';
import { compose } from 'recompose';

import ReportBody from './ReportBody';

const CustomBottomNavigationAction = withStyles((theme) => ({
  root: {
    flex: 0,
    marginLeft: theme.spacing.unit * 3,
  },
}))(BottomNavigationAction);

const propTypes = {
  classes: PropTypes.shape({}).isRequired,
  report: PropTypes.shape({
    cellName: PropTypes.string.isRequired,
    userLabel: PropTypes.string.isRequired,
    lastOccupantName: PropTypes.string,
    startTime: PropTypes.instanceOf(Date).isRequired,
    endTime: PropTypes.instanceOf(Date).isRequired,
    movementHistories: PropTypes.arrayOf(PropTypes.shape({
      time: PropTypes.instanceOf(Date).isRequired,
      detaineeName: PropTypes.string.isRequired,
      sourceCellName: PropTypes.string,
      destinationCellName: PropTypes.string,
    })),
  }),
  isLoadingReport: PropTypes.bool.isRequired,
  history: PropTypes.shape({
    goBack: PropTypes.func.isRequired,
  }).isRequired,
};

const defaultProps = {
  report: null,
};

export class CellHistoryReportComponent extends React.Component {
  handleBackClick = () => {
    const { history } = this.props;
    history.goBack();
  };

  handlePrintClick = () => {
    window.print();
  };

  render() {
    const { classes, report, isLoadingReport } = this.props;
    return (
      <React.Fragment>
        <ReportBody
          isLoadingReport={isLoadingReport}
          report={report}
          ref={(el) => {
            this.componentRef = el;
          }}
        />
        <div className={classNames(classes.footer, 'no-print')}>
          <AppBar position="static" className={classes.appBar}>
            <Grid container alignItems="center">
              <BottomNavigation
                value={-1}
                showLabels
                className={classes.navigation}
              >
                <CustomBottomNavigationAction
                  id="backButton"
                  label={<Typography variant="body1">Back</Typography>}
                  onClick={this.handleBackClick}
                  icon={<ArrowBackIcon className={classes.icon} />}
                  disabled={isLoadingReport}
                  className={classes.backButton}
                />
                <BottomNavigationAction
                  id="printButton"
                  label={<Typography variant="body1">Print</Typography>}
                  icon={<PrintIcon className={classes.icon} />}
                  disabled={isLoadingReport}
                  onClick={this.handlePrintClick}
                  showLabel
                />
              </BottomNavigation>
            </Grid>
          </AppBar>
        </div>
      </React.Fragment>
    );
  }
}

CellHistoryReportComponent.propTypes = propTypes;
CellHistoryReportComponent.defaultProps = defaultProps;

export default compose(
  withStyles((theme) => ({
    footer: {
      flexGrow: 1,
    },
    appBar: {
      backgroundColor: theme.palette.background.default,
    },
    navigation: {
      height: '100%',
      width: '69%',
      backgroundColor: theme.palette.background.default,
      justifyContent: 'start',
    },
    button: {
      height: theme.spacing.unit * 7,
      width: theme.spacing.unit * 40,
      margin: theme.spacing.unit * 2,
    },
    navigationImage: {
      width: theme.spacing.unit * 6,
      height: theme.spacing.unit * 6,
    },
    icon: {
      fontSize: theme.typography.h3.fontSize,
    },
    backButton: {
      marginRight: theme.spacing.unit * 21,
      paddingLeft: 0,
    },
  })),
  withRouter,
)(CellHistoryReportComponent);
