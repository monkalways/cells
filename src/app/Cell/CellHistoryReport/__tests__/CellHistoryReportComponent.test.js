import React from 'react';
import { shallow } from 'enzyme';
import { CellHistoryReportComponent } from '../CellHistoryReportComponent';

describe('CellHistoryReportComponent', () => {
  let isLoadingReport;
  let goBack;

  beforeEach(() => {
    isLoadingReport = false;
    goBack = jest.fn();
  });

  const setup = () => {
    const classes = {};
    return shallow(<CellHistoryReportComponent
      classes={classes}
      report={{
        cellName: 'c1',
        userLabel: 'Doe. J',
        lastOccupantName: 'Test User 1',
        startTime: new Date(Date.UTC(2019, 2, 11, 0)),
        endTime: new Date(Date.UTC(2019, 2, 12, 0)),
        movementHistories: [
          {
            time: new Date(Date.UTC(2019, 2, 11, 0)),
            detaineeName: 'Test User 2',
            sourceCellName: 'c1',
            detinationCellName: 'c2',
          },
        ],
      }}
      isLoadingReport={isLoadingReport}
      history={{ goBack }}
    />);
  };

  it('should render', () => {
    const wrapper = setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('should handleBackClick', () => {
    const wrapper = setup();

    wrapper.find('#backButton').simulate('click');
    expect(goBack).toBeCalled();
  });
});
