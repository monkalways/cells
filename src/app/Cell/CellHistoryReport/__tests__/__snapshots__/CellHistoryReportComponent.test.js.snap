// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CellHistoryReportComponent should render 1`] = `
<Fragment>
  <WithStyles(ReportBodyComponent)
    isLoadingReport={false}
    report={
      Object {
        "cellName": "c1",
        "endTime": 2019-03-12T00:00:00.000Z,
        "lastOccupantName": "Test User 1",
        "movementHistories": Array [
          Object {
            "detaineeName": "Test User 2",
            "detinationCellName": "c2",
            "sourceCellName": "c1",
            "time": 2019-03-11T00:00:00.000Z,
          },
        ],
        "startTime": 2019-03-11T00:00:00.000Z,
        "userLabel": "Doe. J",
      }
    }
  />
  <div
    className="no-print"
  >
    <WithStyles(AppBar)
      position="static"
    >
      <WithStyles(Grid)
        alignItems="center"
        container={true}
      >
        <WithStyles(BottomNavigation)
          showLabels={true}
          value={-1}
        >
          <WithStyles(WithStyles(BottomNavigationAction))
            disabled={false}
            icon={<pure(ArrowBackIcon) />}
            id="backButton"
            label={
              <WithStyles(Typography)
                variant="body1"
              >
                Back
              </WithStyles(Typography)>
            }
            onClick={[Function]}
          />
          <WithStyles(BottomNavigationAction)
            disabled={false}
            icon={<pure(PrintIcon) />}
            id="printButton"
            label={
              <WithStyles(Typography)
                variant="body1"
              >
                Print
              </WithStyles(Typography)>
            }
            onClick={[Function]}
            showLabel={true}
          />
        </WithStyles(BottomNavigation)>
      </WithStyles(Grid)>
    </WithStyles(AppBar)>
  </div>
</Fragment>
`;
